### PR TITLE
Feature/autocomplete

### PR DESCRIPTION
**Pull request by richard-jones**
This adds a mechanism to switch between autocomplete algorithms depending on the incoming string, and supports filtering by one field (e.g. in the index.* document space) and faceting on another, allowing us to more effectively use the prefix query, and to switch seamlessly to the wildcard query.

### CODE REVIEW

**Comment by richard-jones** [comment]
Oh, in order to deploy this the migrate script needs to be run. See

    migrate/autocomplete/README.md

for details

**Comment by Steven-Eardley** [comment]
Looks good to me.

File: portality/migrate/autocomplete/autocomplete.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+	"types": [
+		{
+			"type": "journal"
```
Comment (emanuil-tolev) [inline portality/migrate/autocomplete/autocomplete.json]: Don't applications (type suggestion) also have these auto-complete dropdowns and therefore require a self._generate_autocompletes() to be called on them as well?

Bit unusual to comment on a merged PR - just want to make sure we've done everything consistently.

**Comment by emanuil-tolev** [comment]
:+1: - I had it on my todo to come back to review this properly when I have time just because Linus' law.

**Pull request merged**